Systemtest: Controller getSuccessors API fails with message "No value found"

**Problem description**
In one of the system test runs we find large number of the following exception:

2017-08-02 06:37:55,751 323093 [clientInternal-1] TRACE i.p.c.stream.impl.ControllerImpl - ENTER getSuccessors@11982395992477 [Segment(scope=testReadWriteAndAutoScaleScope1369069020, streamName=testReadWriteAndAutoScaleStream, segmentNumber=5)].
2017-08-02 06:37:55,774 323116 [grpc-default-executor-5] WARN  i.p.c.stream.impl.ControllerImpl - gRPC call failed with server error: INTERNAL: No value present
2017-08-02 06:37:55,775 323117 [grpc-default-executor-5] WARN  i.p.c.stream.impl.ControllerImpl - getSuccessors failed:
java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: INTERNAL: No value present
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593)
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.pravega.client.stream.impl.ControllerImpl$RPCAsyncCallback.onError(ControllerImpl.java:722)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:395)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
        at io.grpc.internal.ClientCallImpl.access$100(ClientCallImpl.java:76)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:512)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:429)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:544)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:52)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:117)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: io.grpc.StatusRuntimeException: INTERNAL: No value present
        at io.grpc.Status.asRuntimeException(Status.java:543)
        ... 11 common frames omitted
2017-08-02 06:37:55,775 323117 [grpc-default-executor-5] TRACE i.p.c.stream.impl.ControllerImpl - LEAVE getSuccessors@11982395992477 (elapsed=23708us).

We could not capture the controller server logs when this happened. "No value present" is typically used by **NoSuchElementException**. So guess is that the API failed in the Controller Service with NoSuchElementException exception.

**Problem location**
Controller's getSuccessors API

**Suggestions for an improvement**
Investigate why the controller API can fail with this error. Since we don't have controller logs for this particular run we need check if this issue reproduces and investigate further.

--- a/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
+++ b/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
@@ -383,7 +383,11 @@ public abstract class PersistentStreamBase<T> implements Stream {
                         // fresh run
                         // Ensure that segment.creation time is monotonically increasing after each new scale
                         long lastScaleTime = HistoryRecord.readLatestRecord(historyTable.getData(), true).map(HistoryRecord::getScaleTime).orElse(0L);
-                        long segmentCreationTimestamp = Math.max(scaleTimestamp, lastScaleTime + 1);
+
+                        long scaleEventTime = Math.max(System.currentTimeMillis(), scaleTimestamp);
+
+                        long segmentCreationTimestamp = Math.max(scaleEventTime, lastScaleTime + 1);
+
                         return scaleCreateNewSegments(newRanges, segmentCreationTimestamp, segmentTable, activeEpoch);
                     }
                 })

--- a/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
+++ b/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
@@ -381,7 +381,10 @@ public abstract class PersistentStreamBase<T> implements Stream {
 
                         log.info("Scale {}/{} for segments started. Creating new segments. SegmentsToSeal {}", scope, name, sealedSegments);
                         // fresh run
-                        return scaleCreateNewSegments(newRanges, scaleTimestamp, segmentTable, activeEpoch);
+                        // Ensure that segment.creation time is monotonically increasing after each new scale
+                        long lastScaleTime = HistoryRecord.readLatestRecord(historyTable.getData(), true).map(HistoryRecord::getScaleTime).orElse(0L);
+                        long segmentCreationTimestamp = Math.max(scaleTimestamp, lastScaleTime + 1);
+                        return scaleCreateNewSegments(newRanges, segmentCreationTimestamp, segmentTable, activeEpoch);
                     }
                 })
                 .thenCompose(epochStartSegmentpair -> getSegments(IntStream.range(epochStartSegmentpair.getRight(),

--- a/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
+++ b/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
@@ -241,8 +241,8 @@ public abstract class PersistentStreamBase<T> implements Stream {
     private CompletableFuture<List<Segment>> getSuccessorsForSegment(final int number) {
         return getHistoryTable()
                 .thenApply(historyTable -> {
-                    val segmentFuture = getSegment(number);
-                    val indexTableFuture = getIndexTable();
+                    CompletableFuture<Segment> segmentFuture = getSegment(number);
+                    CompletableFuture<Data<T>> indexTableFuture = getIndexTable();
                     return new ImmutableTriple<>(historyTable, segmentFuture, indexTableFuture);
                 })
                 .thenCompose(triple -> CompletableFuture.allOf(triple.getMiddle(), triple.getRight())

--- a/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
+++ b/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
@@ -27,9 +27,8 @@ import lombok.extern.slf4j.Slf4j;
 import lombok.val;
 import org.apache.commons.lang.SerializationUtils;
 import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.ImmutableTriple;
 import org.apache.commons.lang3.tuple.Pair;
-
-import static java.util.stream.Collectors.toMap;
 
 import java.util.AbstractMap;
 import java.util.AbstractMap.SimpleEntry;
@@ -44,6 +43,8 @@ import java.util.concurrent.CompletionException;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+
+import static java.util.stream.Collectors.toMap;
 
 @Slf4j
 public abstract class PersistentStreamBase<T> implements Stream {
@@ -238,41 +239,48 @@ public abstract class PersistentStreamBase<T> implements Stream {
     }
 
     private CompletableFuture<List<Segment>> getSuccessorsForSegment(final int number) {
-        val segmentFuture = getSegment(number);
-        val indexTableFuture = getIndexTable();
-        val historyTableFuture = getHistoryTable();
-        CompletableFuture<Void> all = CompletableFuture.allOf(segmentFuture, indexTableFuture, historyTableFuture);
-
-        return all.thenCompose(x -> {
-            final Segment segment = segmentFuture.getNow(null);
-            List<Integer> candidates = TableHelper.findSegmentSuccessorCandidates(segment,
-                    indexTableFuture.getNow(null).getData(),
-                    historyTableFuture.getNow(null).getData());
-            return findOverlapping(segment, candidates);
-        });
+        return getHistoryTable()
+                .thenApply(historyTable -> {
+                    val segmentFuture = getSegment(number);
+                    val indexTableFuture = getIndexTable();
+                    return new ImmutableTriple<>(historyTable, segmentFuture, indexTableFuture);
+                })
+                .thenCompose(triple -> CompletableFuture.allOf(triple.getMiddle(), triple.getRight())
+                        .thenCompose(x -> {
+                            final Segment segment = triple.getMiddle().join();
+                            List<Integer> candidates = TableHelper.findSegmentSuccessorCandidates(segment,
+                                    triple.getRight().join().getData(),
+                                    triple.getLeft().getData());
+                            return findOverlapping(segment, candidates);
+                        }));
     }
 
     @Override
     public CompletableFuture<Map<Integer, List<Integer>>> getSuccessorsWithPredecessors(final int number) {
-        val legal = verifyLegalState();
-        val indexTableFuture = getIndexTable();
-        val historyTableFuture = getHistoryTable();
-        CompletableFuture<List<Segment>> segments = getSuccessorsForSegment(number);
+        // Ensure the order, we should first get history table followed by segment table because during scale we first write to
+        // segment table followed by history table. So if a record exists in history table, then we are guaranteed to find it in
+        // segment table.
+        return verifyLegalState()
+                .thenCompose(legal -> getHistoryTable()
+                        .thenApply(historyTable -> new ImmutableTriple<>(historyTable, getIndexTable(), getSuccessorsForSegment(number)))
+                        .thenCompose(triple -> CompletableFuture.allOf(triple.getMiddle(), triple.getRight())
+                                .thenCompose(x -> {
+                                    List<CompletableFuture<Map.Entry<Segment, List<Integer>>>> resultFutures = new ArrayList<>();
+                                    List<Segment> successors = triple.getRight().join();
+                                    Data<T> indexTable = triple.getMiddle().join();
+                                    Data<T> historyTable = triple.getLeft();
 
-        CompletableFuture<Void> all = CompletableFuture.allOf(legal, segments, indexTableFuture, historyTableFuture);
-
-        return all.thenCompose(v -> {
-            List<CompletableFuture<Map.Entry<Segment, List<Integer>>>> resultFutures = new ArrayList<>();
-            List<Segment> successors = segments.getNow(null);
-            for (Segment successor : successors) {
-                List<Integer> candidates = TableHelper.findSegmentPredecessorCandidates(successor,
-                        indexTableFuture.getNow(null).getData(),
-                        historyTableFuture.getNow(null).getData());
-                resultFutures.add(findOverlapping(successor, candidates).thenApply(
-                        list -> new SimpleImmutableEntry<>(successor, list.stream().map(Segment::getNumber).collect(Collectors.toList()))));
-            }
-            return FutureHelpers.allOfWithResults(resultFutures);
-        }).thenApply(list -> list.stream().collect(Collectors.toMap(e -> e.getKey().getNumber(), e -> e.getValue())));
+                                    for (Segment successor : successors) {
+                                        List<Integer> candidates = TableHelper.findSegmentPredecessorCandidates(successor,
+                                                indexTable.getData(),
+                                                historyTable.getData());
+                                        resultFutures.add(findOverlapping(successor, candidates).thenApply(
+                                                list -> new SimpleImmutableEntry<>(successor, list.stream().map(Segment::getNumber)
+                                                        .collect(Collectors.toList()))));
+                                    }
+                                    return FutureHelpers.allOfWithResults(resultFutures);
+                                })
+                                .thenApply(list -> list.stream().collect(Collectors.toMap(e -> e.getKey().getNumber(), Map.Entry::getValue)))));
     }
 
     /**
@@ -283,19 +291,17 @@ public abstract class PersistentStreamBase<T> implements Stream {
      */
     @Override
     public CompletableFuture<List<Integer>> getPredecessors(final int number) {
-        val legal = verifyLegalState();
-        val segmentFuture = getSegment(number);
-        val indexTableFuture = getIndexTable();
-        val historyTableFuture = getHistoryTable();
-        CompletableFuture<Void> all = CompletableFuture.allOf(legal, segmentFuture, indexTableFuture, historyTableFuture);
-
-        return all.thenCompose(x -> {
-            final Segment segment = segmentFuture.getNow(null);
-            List<Integer> candidates = TableHelper.findSegmentPredecessorCandidates(segment,
-                    indexTableFuture.getNow(null).getData(),
-                    historyTableFuture.getNow(null).getData());
-            return findOverlapping(segment, candidates);
-        }).thenApply(list -> list.stream().map(e -> e.getNumber()).collect(Collectors.toList()));
+        return verifyLegalState().thenCompose(x -> getHistoryTable()
+                .thenApply(historyTable -> new ImmutableTriple<>(historyTable, getSegment(number), getIndexTable())))
+                .thenCompose(triple -> CompletableFuture.allOf(triple.getMiddle(), triple.getRight()).thenCompose(x -> {
+                    final Segment segment = triple.getMiddle().join();
+                    Data<T> indexTable = triple.getRight().join();
+                    Data<T> historyTable = triple.getLeft();
+                    List<Integer> candidates = TableHelper.findSegmentPredecessorCandidates(segment,
+                            indexTable.getData(),
+                            historyTable.getData());
+                    return findOverlapping(segment, candidates);
+                }).thenApply(list -> list.stream().map(Segment::getNumber).collect(Collectors.toList())));
     }
 
     @Override

--- a/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
+++ b/controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java
@@ -381,7 +381,9 @@ public abstract class PersistentStreamBase<T> implements Stream {
 
                         log.info("Scale {}/{} for segments started. Creating new segments. SegmentsToSeal {}", scope, name, sealedSegments);
                         // fresh run
-                        // Ensure that segment.creation time is monotonically increasing after each new scale
+                        // Ensure that segment.creation time is monotonically increasing after each new scale.
+                        // because scale time could be supplied by a controller with a skewed clock, we should:
+                        // take max(scaleTime, lastScaleTime + 1, System.currentTimeMillis)
                         long lastScaleTime = HistoryRecord.readLatestRecord(historyTable.getData(), true).map(HistoryRecord::getScaleTime).orElse(0L);
 
                         long scaleEventTime = Math.max(System.currentTimeMillis(), scaleTimestamp);

--- a/controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java
+++ b/controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java
@@ -164,15 +164,16 @@ public class TableHelper {
         final Optional<IndexRecord> recordOpt = search.getValue();
         final int startingOffset = recordOpt.isPresent() ? recordOpt.get().getHistoryOffset() : 0;
 
-        final Optional<HistoryRecord> historyRecordOpt = findSegmentCreatedEvent(startingOffset,
+        final Optional<HistoryRecord> segmentCreatedHistoryRecordOpt = findSegmentCreatedEvent(startingOffset,
                 segment, historyTable);
 
         // segment information not in history table
-        if (!historyRecordOpt.isPresent()) {
+        if (!segmentCreatedHistoryRecordOpt.isPresent()) {
             return new ArrayList<>();
         }
 
-        final int lower = search.getKey() / IndexRecord.INDEX_RECORD_SIZE;
+        // take index of segment created event instead of searched index based on segment.startTime.
+        final int lower = IndexRecord.search(segmentCreatedHistoryRecordOpt.get().getScaleTime(), indexTable).getKey() / IndexRecord.INDEX_RECORD_SIZE;
 
         final int upper = (indexTable.length - IndexRecord.INDEX_RECORD_SIZE) / IndexRecord.INDEX_RECORD_SIZE;
 
@@ -187,11 +188,12 @@ public class TableHelper {
         // we cant do anything on index table, fall through. OR
         // if segment exists at the last indexed record in history table, fall through,
         // no binary search possible on index
-        if (lastIndexedRecord.get().getScaleTime() < historyRecordOpt.get().getScaleTime() ||
+        // Note: lower will always be lessThanEq upper. So if upper.getScaleTime < segmentCreatedRecord.ScaleTime then lower < segmentCreatedRecord.ScaleTime.
+        if (lastIndexedRecord.get().getScaleTime() < segmentCreatedHistoryRecordOpt.get().getScaleTime() ||
                 lastIndexedRecord.get().getSegments().contains(segment.getNumber())) {
             // segment was sealed after the last index entry
-            HistoryRecord startPoint = lastIndexedRecord.get().getScaleTime() < historyRecordOpt.get().getScaleTime() ?
-                    historyRecordOpt.get() : lastIndexedRecord.get();
+            HistoryRecord startPoint = lastIndexedRecord.get().getScaleTime() < segmentCreatedHistoryRecordOpt.get().getScaleTime() ?
+                    segmentCreatedHistoryRecordOpt.get() : lastIndexedRecord.get();
             Optional<HistoryRecord> next = HistoryRecord.fetchNext(startPoint, historyTable, false);
 
             while (next.isPresent() && next.get().getSegments().contains(segment.getNumber())) {
@@ -572,7 +574,15 @@ public class TableHelper {
         return Optional.of(record);
     }
 
-
+    /**
+     * It finds the segment sealed event between lower and upper where 'lower' offset is guaranteed to be greater than or equal to segmentCreatedEvent
+     * @param lower starting record number in index table from where to search
+     * @param upper last record number in index table till where to search
+     * @param segmentNumber segment number to find sealed event
+     * @param indexTable index table
+     * @param historyTable history table
+     * @return returns history record where segment was sealed
+     */
     private static Optional<HistoryRecord> findSegmentSealedEvent(final int lower,
                                                                   final int upper,
                                                                   final int segmentNumber,

--- a/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
@@ -141,7 +141,7 @@ public class StreamTest {
     }
 
     @Test
-    public void testNoValuePresent()  throws Exception {
+    public void testNoValuePresent() throws Exception {
         final ScalingPolicy policy = ScalingPolicy.fixed(1);
 
         final StreamMetadataStore store = new ZKStreamMetadataStore(cli, executor);
@@ -186,7 +186,7 @@ public class StreamTest {
         AtomicBoolean segmentCalled = new AtomicBoolean(false);
         // mock.. If segment table is fetched before history table, throw runtime exception so that the test fails
         doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
-            if (segmentCalled.get()) {
+            if (!historyCalled.get() && segmentCalled.get()) {
                 throw new RuntimeException();
             }
             historyCalled.set(true);
@@ -217,7 +217,7 @@ public class StreamTest {
         segmentCalled.set(false);
         historyCalled.set(false);
         doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
-            if (segmentCalled.get()) {
+            if (!historyCalled.get() && segmentCalled.get()) {
                 throw new RuntimeException();
             }
             historyCalled.set(true);

--- a/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
@@ -193,7 +193,7 @@ public class StreamTest {
             return historyTable;
         }).when(zkStream).getHistoryTable();
         doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
-            if (!historyCalled.get()){
+            if (!historyCalled.get()) {
                 throw new RuntimeException();
             }
             segmentCalled.set(true);
@@ -224,7 +224,7 @@ public class StreamTest {
             return historyTable;
         }).when(zkStream).getHistoryTable();
         doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
-            if (!historyCalled.get()){
+            if (!historyCalled.get()) {
                 throw new RuntimeException();
             }
             segmentCalled.set(true);

--- a/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
@@ -12,6 +12,7 @@ package io.pravega.controller.store.stream;
 import com.google.common.collect.Lists;
 import io.pravega.client.stream.ScalingPolicy;
 import io.pravega.client.stream.StreamConfiguration;
+import io.pravega.controller.store.stream.tables.Data;
 import io.pravega.controller.store.stream.tables.State;
 import io.pravega.test.common.TestingServerStarter;
 import org.apache.curator.framework.CuratorFramework;
@@ -21,18 +22,24 @@ import org.apache.curator.test.TestingServer;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.stubbing.Answer;
 
 import java.util.AbstractMap;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.spy;
 
 public class StreamTest {
     private TestingServer zkTestServer;
@@ -153,7 +160,7 @@ public class StreamTest {
         store.createStream(scopeName, streamName, streamConfig, System.currentTimeMillis(), null, executor).get();
         store.setState(scopeName, streamName, State.ACTIVE, null, executor).get();
 
-        ZKStream zkStream = new ZKStream("test", "test", zkStoreHelper);
+        ZKStream zkStream = spy(new ZKStream("test", "test", zkStoreHelper));
 
         List<AbstractMap.SimpleEntry<Double, Double>> newRanges;
 
@@ -171,12 +178,61 @@ public class StreamTest {
                 response.getActiveEpoch(), scale).get();
         // history table has a partial record at this point.
         // now we could have sealed the segments so get successors could be called.
+
+        final CompletableFuture<Data<Integer>> segmentTable = zkStream.getSegmentTable();
+        final CompletableFuture<Data<Integer>> historyTable = zkStream.getHistoryTable();
+
+        AtomicBoolean historyCalled = new AtomicBoolean(false);
+        AtomicBoolean segmentCalled = new AtomicBoolean(false);
+        // mock.. If segment table is fetched before history table, throw runtime exception so that the test fails
+        doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
+            if (segmentCalled.get()) {
+                throw new RuntimeException();
+            }
+            historyCalled.set(true);
+            return historyTable;
+        }).when(zkStream).getHistoryTable();
+        doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
+            if (!historyCalled.get()){
+                throw new RuntimeException();
+            }
+            segmentCalled.set(true);
+            return segmentTable;
+        }).when(zkStream).getSegmentTable();
+
         Map<Integer, List<Integer>> successors = zkStream.getSuccessorsWithPredecessors(0).get();
+
+        assertTrue(successors.containsKey(1) && successors.containsKey(2));
+
+        // reset mock so that we can resume scale operation
+        doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> historyTable).when(zkStream).getHistoryTable();
+        doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> segmentTable).when(zkStream).getSegmentTable();
 
         zkStream.scaleOldSegmentsSealed(sealedSegments, newSegmentInt, response.getActiveEpoch(), scale).get();
         // scale is completed, history table also has completed record now.
+        final CompletableFuture<Data<Integer>> segmentTable2 = zkStream.getSegmentTable();
+        final CompletableFuture<Data<Integer>> historyTable2 = zkStream.getHistoryTable();
+
+        // mock such that if segment table is fetched before history table, throw runtime exception so that the test fails
+        segmentCalled.set(false);
+        historyCalled.set(false);
+        doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
+            if (segmentCalled.get()) {
+                throw new RuntimeException();
+            }
+            historyCalled.set(true);
+            return historyTable;
+        }).when(zkStream).getHistoryTable();
+        doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
+            if (!historyCalled.get()){
+                throw new RuntimeException();
+            }
+            segmentCalled.set(true);
+            return segmentTable;
+        }).when(zkStream).getSegmentTable();
 
         successors = zkStream.getSuccessorsWithPredecessors(0).get();
-    }
 
+        assertTrue(successors.containsKey(1) && successors.containsKey(2));
+    }
 }

--- a/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
@@ -140,8 +140,8 @@ public class StreamTest {
         assertEquals(CreateStreamResponse.CreateStatus.EXISTS_ACTIVE, response.getStatus());
     }
 
-    @Test
-    public void testNoValuePresent() throws Exception {
+    @Test(timeout = 10000)
+    public void testConcurrentGetSuccessorScale() throws Exception {
         final ScalingPolicy policy = ScalingPolicy.fixed(1);
 
         final StreamMetadataStore store = new ZKStreamMetadataStore(cli, executor);
@@ -221,14 +221,14 @@ public class StreamTest {
                 throw new RuntimeException();
             }
             historyCalled.set(true);
-            return historyTable;
+            return historyTable2;
         }).when(zkStream).getHistoryTable();
         doAnswer((Answer<CompletableFuture<Data<Integer>>>) invocation -> {
             if (!historyCalled.get()) {
                 throw new RuntimeException();
             }
             segmentCalled.set(true);
-            return segmentTable;
+            return segmentTable2;
         }).when(zkStream).getSegmentTable();
 
         successors = zkStream.getSuccessorsWithPredecessors(0).get();

--- a/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/StreamTest.java
@@ -9,6 +9,7 @@
  */
 package io.pravega.controller.store.stream;
 
+import com.google.common.collect.Lists;
 import io.pravega.client.stream.ScalingPolicy;
 import io.pravega.client.stream.StreamConfiguration;
 import io.pravega.controller.store.stream.tables.State;
@@ -21,9 +22,15 @@ import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.stream.Collectors;
 
 import static org.junit.Assert.assertEquals;
 
@@ -124,6 +131,52 @@ public class StreamTest {
         assertEquals(CreateStreamResponse.CreateStatus.EXISTS_ACTIVE, response.getStatus());
         response = stream.checkStreamExists(streamConfig2, creationTime2).get();
         assertEquals(CreateStreamResponse.CreateStatus.EXISTS_ACTIVE, response.getStatus());
+    }
+
+    @Test
+    public void testNoValuePresent()  throws Exception {
+        final ScalingPolicy policy = ScalingPolicy.fixed(1);
+
+        final StreamMetadataStore store = new ZKStreamMetadataStore(cli, executor);
+        final String streamName = "test";
+        String scopeName = "test";
+        store.createScope(scopeName).get();
+
+        ZKStoreHelper zkStoreHelper = new ZKStoreHelper(cli, executor);
+
+        StreamConfiguration streamConfig = StreamConfiguration.builder()
+                .scope(streamName)
+                .streamName(streamName)
+                .scalingPolicy(policy)
+                .build();
+
+        store.createStream(scopeName, streamName, streamConfig, System.currentTimeMillis(), null, executor).get();
+        store.setState(scopeName, streamName, State.ACTIVE, null, executor).get();
+
+        ZKStream zkStream = new ZKStream("test", "test", zkStoreHelper);
+
+        List<AbstractMap.SimpleEntry<Double, Double>> newRanges;
+
+        newRanges = Arrays.asList(new AbstractMap.SimpleEntry<>(0.0, 0.5), new AbstractMap.SimpleEntry<>(0.5, 1.0));
+
+        long scale = System.currentTimeMillis();
+        ArrayList<Integer> sealedSegments = Lists.newArrayList(0);
+
+        StartScaleResponse response = zkStream.startScale(sealedSegments, newRanges, scale, false).join();
+        List<Segment> newSegments = response.getSegmentsCreated();
+        zkStream.updateState(State.SCALING).join();
+
+        List<Integer> newSegmentInt = newSegments.stream().map(Segment::getNumber).collect(Collectors.toList());
+        zkStream.scaleNewSegmentsCreated(sealedSegments, newSegmentInt,
+                response.getActiveEpoch(), scale).get();
+        // history table has a partial record at this point.
+        // now we could have sealed the segments so get successors could be called.
+        Map<Integer, List<Integer>> successors = zkStream.getSuccessorsWithPredecessors(0).get();
+
+        zkStream.scaleOldSegmentsSealed(sealedSegments, newSegmentInt, response.getActiveEpoch(), scale).get();
+        // scale is completed, history table also has completed record now.
+
+        successors = zkStream.getSuccessorsWithPredecessors(0).get();
     }
 
 }

--- a/controller/src/test/java/io/pravega/controller/store/stream/TableHelperTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/TableHelperTest.java
@@ -16,8 +16,10 @@ import io.pravega.controller.store.stream.tables.TableHelper;
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.text.ParseException;
 import java.util.AbstractMap;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
@@ -168,6 +170,162 @@ public class TableHelperTest {
                         .collect(Collectors.toList()));
         assertEquals(predecessors, Lists.newArrayList(2, 3, 4));
         assertEquals(successors, Lists.newArrayList(6));
+    }
+
+    @Test
+    public void testSegmentCreationBeforePreviousScale() throws ParseException {
+        // no value present error comes because:
+        // - Index is not yet updated.
+        // - And segment creation time is before history record's time.
+        // While trying to find successor we look for record in history table with
+        // segment creation and get an old record. We search for segment sealed event
+        // between history record and last indexed entry both of which preceed segment creation entry.
+        List<Segment> segments = new ArrayList<>();
+        List<Integer> newSegments = Lists.newArrayList(0, 1);
+        // create stream
+        long timestamp = 1503933145366L;
+        Segment zero = new Segment(0, timestamp, 0, 0.5);
+        segments.add(zero);
+        Segment one = new Segment(1, timestamp, 0.5, 1.0);
+        segments.add(one);
+
+        byte[] historyTable = TableHelper.createHistoryTable(timestamp, newSegments);
+        byte[] indexTable = TableHelper.createIndexTable(timestamp, 0);
+
+        // scale up 1... 0 -> 2, 3
+        int numOfSplits = 2;
+        double delta = (zero.getKeyEnd() - zero.getKeyStart()) / numOfSplits;
+
+        ArrayList<AbstractMap.SimpleEntry<Double, Double>> simpleEntries = new ArrayList<>();
+        for (int i = 0; i < numOfSplits; i++) {
+            simpleEntries.add(new AbstractMap.SimpleEntry<>(zero.getKeyStart() + delta * i,
+                    zero.getKeyStart() + (delta * (i + 1))));
+        }
+
+        Segment two = new Segment(2, 1503933266113L, simpleEntries.get(0).getKey(), simpleEntries.get(0).getValue());
+        segments.add(two);
+        Segment three = new Segment(3, 1503933266113L, simpleEntries.get(1).getKey(), simpleEntries.get(1).getValue());
+        segments.add(three);
+
+        newSegments = Lists.newArrayList(1, 2, 3);
+
+        historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
+
+        HistoryRecord partial = HistoryRecord.readLatestRecord(historyTable, false).get();
+        // Notice: segment was created at timestamp but we are recording its entry in history table at timestamp + 10000
+        timestamp = 1503933266862L;
+
+        historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
+        HistoryRecord historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
+
+        indexTable = TableHelper.updateIndexTable(indexTable,
+                historyRecord.getScaleTime(),
+                historyRecord.getOffset());
+
+        // scale up 2.. 1 -> 4, 5
+        delta = (one.getKeyEnd() - one.getKeyStart()) / numOfSplits;
+
+        simpleEntries = new ArrayList<>();
+        for (int i = 0; i < numOfSplits; i++) {
+            simpleEntries.add(new AbstractMap.SimpleEntry<>(one.getKeyStart() + delta * i,
+                    one.getKeyStart() + (delta * (i + 1))));
+        }
+
+        Segment four = new Segment(4, 1503933266188L, simpleEntries.get(0).getKey(), simpleEntries.get(0).getValue());
+        segments.add(four);
+
+        Segment five = new Segment(5, 1503933266188L, simpleEntries.get(1).getKey(), simpleEntries.get(1).getValue());
+        segments.add(five);
+
+        newSegments = Lists.newArrayList(2, 3, 4, 5);
+
+        historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
+
+        partial = HistoryRecord.readLatestRecord(historyTable, false).get();
+        // Notice: segment was created at timestamp but we are recording its entry in history table at timestamp + 10000
+        timestamp = 1503933288726L;
+
+        historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
+        historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
+
+        indexTable = TableHelper.updateIndexTable(indexTable,
+                historyRecord.getScaleTime(),
+                historyRecord.getOffset());
+
+        // scale up 3.. 5 -> 6, 7
+        delta = (five.getKeyEnd() - five.getKeyStart()) / numOfSplits;
+
+        simpleEntries = new ArrayList<>();
+        for (int i = 0; i < numOfSplits; i++) {
+            simpleEntries.add(new AbstractMap.SimpleEntry<>(five.getKeyStart() + delta * i,
+                    five.getKeyStart() + (delta * (i + 1))));
+        }
+
+        Segment six = new Segment(6, 1503933409076L, simpleEntries.get(0).getKey(), simpleEntries.get(0).getValue());
+        segments.add(six);
+
+        Segment seven = new Segment(7, 1503933409076L, simpleEntries.get(1).getKey(), simpleEntries.get(1).getValue());
+        segments.add(seven);
+
+        newSegments = Lists.newArrayList(2, 3, 4, 6, 7);
+
+        historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
+
+        partial = HistoryRecord.readLatestRecord(historyTable, false).get();
+
+        timestamp = 1503933409806L;
+
+        List<Integer> candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
+                indexTable,
+                historyTable);
+
+        assertTrue(candidates5.containsAll(Arrays.asList(2, 3, 4, 6, 7)));
+
+        historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
+        historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
+
+        candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
+                indexTable,
+                historyTable);
+        assertTrue(candidates5.containsAll(Arrays.asList(2, 3, 4, 6, 7)));
+
+        indexTable = TableHelper.updateIndexTable(indexTable,
+                historyRecord.getScaleTime(),
+                historyRecord.getOffset());
+
+        candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
+                indexTable,
+                historyTable);
+        assertTrue(candidates5.containsAll(Arrays.asList(2, 3, 4, 6, 7)));
+
+        // scale down 6, 7 -> 8
+
+        // scale down
+        timestamp = 1503933560447L;
+
+        Segment eight = new Segment(8, timestamp, six.keyStart, seven.keyEnd);
+        segments.add(eight);
+
+        newSegments = Lists.newArrayList(2, 3, 4, 8);
+
+        historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
+
+        partial = HistoryRecord.readLatestRecord(historyTable, false).get();
+        // Notice: segment was created at timestamp but we are recording its entry in history table at timestamp + 10000
+        timestamp = 1503933560448L;
+
+        historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
+        historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
+
+        indexTable = TableHelper.updateIndexTable(indexTable,
+                historyRecord.getScaleTime(),
+                historyRecord.getOffset());
+
+        candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
+                indexTable,
+                historyTable);
+
+        assertTrue(candidates5.containsAll(Arrays.asList(2, 3, 4, 6, 7)));
     }
 
     @Test

--- a/controller/src/test/java/io/pravega/controller/store/stream/TableHelperTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/TableHelperTest.java
@@ -172,7 +172,7 @@ public class TableHelperTest {
         assertEquals(successors, Lists.newArrayList(6));
     }
 
-    @Test
+    @Test(timeout = 10000)
     public void testSegmentCreationBeforePreviousScale() throws ParseException {
         // no value present error comes because:
         // - Index is not yet updated.

--- a/controller/src/test/java/io/pravega/controller/store/stream/TableHelperTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/TableHelperTest.java
@@ -202,6 +202,7 @@ public class TableHelperTest {
                     zero.getKeyStart() + (delta * (i + 1))));
         }
 
+        // create segments before scale
         Segment two = new Segment(2, 1503933266113L, simpleEntries.get(0).getKey(), simpleEntries.get(0).getValue());
         segments.add(two);
         Segment three = new Segment(3, 1503933266113L, simpleEntries.get(1).getKey(), simpleEntries.get(1).getValue());
@@ -209,12 +210,14 @@ public class TableHelperTest {
 
         newSegments = Lists.newArrayList(1, 2, 3);
 
+        // add partial record to history table
         historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
 
         HistoryRecord partial = HistoryRecord.readLatestRecord(historyTable, false).get();
-        // Notice: segment was created at timestamp but we are recording its entry in history table at timestamp + 10000
+
         timestamp = 1503933266862L;
 
+        // complete record in history table by adding time
         historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
         HistoryRecord historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
 
@@ -230,7 +233,7 @@ public class TableHelperTest {
             simpleEntries.add(new AbstractMap.SimpleEntry<>(one.getKeyStart() + delta * i,
                     one.getKeyStart() + (delta * (i + 1))));
         }
-
+        // create segments before scale
         Segment four = new Segment(4, 1503933266188L, simpleEntries.get(0).getKey(), simpleEntries.get(0).getValue());
         segments.add(four);
 
@@ -239,12 +242,14 @@ public class TableHelperTest {
 
         newSegments = Lists.newArrayList(2, 3, 4, 5);
 
+        // add partial record to history table
         historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
 
         partial = HistoryRecord.readLatestRecord(historyTable, false).get();
         // Notice: segment was created at timestamp but we are recording its entry in history table at timestamp + 10000
         timestamp = 1503933288726L;
 
+        // complete record in history table by adding time
         historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
         historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
 
@@ -261,6 +266,7 @@ public class TableHelperTest {
                     five.getKeyStart() + (delta * (i + 1))));
         }
 
+        // create new segments
         Segment six = new Segment(6, 1503933409076L, simpleEntries.get(0).getKey(), simpleEntries.get(0).getValue());
         segments.add(six);
 
@@ -268,22 +274,23 @@ public class TableHelperTest {
         segments.add(seven);
 
         newSegments = Lists.newArrayList(2, 3, 4, 6, 7);
-
+        // create partial record in history table
         historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
 
         partial = HistoryRecord.readLatestRecord(historyTable, false).get();
 
         timestamp = 1503933409806L;
-
+        // find successor candidates before completing scale.
         List<Integer> candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
                 indexTable,
                 historyTable);
 
         assertTrue(candidates5.containsAll(Arrays.asList(2, 3, 4, 6, 7)));
-
+        // complete record in history table by adding time
         historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
         historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
 
+        // verify successor candidates after completing history record but before adding index entry
         candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
                 indexTable,
                 historyTable);
@@ -293,6 +300,7 @@ public class TableHelperTest {
                 historyRecord.getScaleTime(),
                 historyRecord.getOffset());
 
+        // verify successor candidates after index is updated
         candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
                 indexTable,
                 historyTable);
@@ -302,7 +310,7 @@ public class TableHelperTest {
 
         // scale down
         timestamp = 1503933560447L;
-
+        // add another scale
         Segment eight = new Segment(8, timestamp, six.keyStart, seven.keyEnd);
         segments.add(eight);
 
@@ -311,16 +319,16 @@ public class TableHelperTest {
         historyTable = TableHelper.addPartialRecordToHistoryTable(historyTable, newSegments);
 
         partial = HistoryRecord.readLatestRecord(historyTable, false).get();
-        // Notice: segment was created at timestamp but we are recording its entry in history table at timestamp + 10000
         timestamp = 1503933560448L;
-
+        // complete scale
         historyTable = TableHelper.completePartialRecordInHistoryTable(historyTable, partial, timestamp);
         historyRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
-
+        // add index
         indexTable = TableHelper.updateIndexTable(indexTable,
                 historyRecord.getScaleTime(),
                 historyRecord.getOffset());
 
+        // verify successors again after a new scale entry comes in
         candidates5 = TableHelper.findSegmentSuccessorCandidates(five,
                 indexTable,
                 historyTable);

